lucida/imagematching/opencv_imm/server/Image.cpp

```diff
@@ -31,220 +31,223 @@ namespace po = boost::program_options;
-	rtn->convertTo(*rtn, OPENCV_TYPE);
-	return rtn;
+   unique_ptr<Mat> rtn(new Mat());
+   stringstream ss(mat);
+   for (string line; getline(ss, line); ) {
+      vector<double> dvals;
+      stringstream ssline(line);
+      for (string val; getline(ssline, val, ','); )  {
+         dvals.push_back(stod(val));
+      }
+      Mat mline(dvals, true);
+      transpose(mline, mline);
+      rtn->push_back(mline);
+   }
+   int ch = CV_MAT_CN(OPENCV_TYPE);
+   *rtn = rtn->reshape(ch);
+   rtn->convertTo(*rtn, OPENCV_TYPE);
+   return rtn;
 }
 
 int Image::match(
-		vector<unique_ptr<StoredImage>> &train_images,
-		unique_ptr<QueryImage> query_image) {
-	if (train_images.empty()) {
-		throw runtime_error("Error! No images!");
-	}
-	// Train a FlannBasedMatcher.
-	unique_ptr<DescriptorMatcher> matcher(new FlannBasedMatcher());
-	vector<Mat> train_mats;
-	for (unique_ptr<StoredImage> &image_ptr : train_images) {
-		train_mats.push_back(*(image_ptr->desc));
-	}
-	matcher->add(train_mats);
-	matcher->train();
-	// Prepare to match.
-	vector<vector<DMatch>> knn_matches;
-	vector<int> scores(train_images.size(), 0);
-	int knn = 1;
-	// Match.
-	matcher->knnMatch(*(query_image->desc), knn_matches, knn);
-	// Filter results.
-	for (auto &v : knn_matches) {
-		for(auto &dMatch : v){
-			++scores[dMatch.imgIdx];
-		}
-	}
-	// Find the best match.
-	auto best = max_element(scores.begin(), scores.end());
-	return best - scores.begin();
+      vector<unique_ptr<StoredImage>> &train_images,
+      unique_ptr<QueryImage> query_image) {
+   if (train_images.empty()) {
+      throw runtime_error("Error! No images!");
+   }
+   // Train a FlannBasedMatcher.
+   unique_ptr<DescriptorMatcher> matcher(new FlannBasedMatcher());
+   vector<Mat> train_mats;
+   for (unique_ptr<StoredImage> &image_ptr : train_images) {
+       if(image_ptr->desc->type() != OPENCV_TYPE) {
+           image_ptr->desc->convertTo(*(image_ptr->desc), OPENCV_TYPE);
+       }
+       train_mats.push_back(*(image_ptr->desc));
+   }
+   matcher->add(train_mats);
+   matcher->train();
+   // Prepare to match.
+   vector<vector<DMatch>> knn_matches;
+   vector<int> scores(train_images.size(), 0);
+   int knn = 1;
+   // Match.
+   matcher->knnMatch(*(query_image->desc), knn_matches, knn);
+   // Filter results.
+   for (auto &v : knn_matches) {
+      for(auto &dMatch : v){
+         ++scores[dMatch.imgIdx];
+      }
+   }
+   // Find the best match.
+   auto best = max_element(scores.begin(), scores.end());
+   return best - scores.begin();
 }
 
 bool Image::matEqual(unique_ptr<Mat> a, unique_ptr<Mat> b) {
-	return matToVector(move(a)) == matToVector(move(b));
+   return matToVector(move(a)) == matToVector(move(b));
 }
 
 //struct timeval tv1, tv2;
 //int debug = 0;
 
 //string make_pbdesc(string img_name) {
-//	string res;
-//	size_t pos = img_name.find(".jpg");
-//	if (pos != std::string::npos) res = img_name.substr(0, pos);
-//	res = res + ".pb";
-//	return res;
+//   string res;
```

Comment:
what's all this commented out code? maybe not you but if you know put a comment/